agent/rpc/middleware/interceptors_test.go

```diff
@@ -57,75 +59,195 @@ func (wr writeRequest) IsRead() bool {
 	return false
 }
 
-// TestRequestRecorder_SimpleOK tests that the RequestRecorder can record a simple request.
-func TestRequestRecorder_SimpleOK(t *testing.T) {
-	t.Parallel()
-
-	r := RequestRecorder{
-		Logger:       hclog.NewInterceptLogger(&hclog.LoggerOptions{}),
-		RecorderFunc: simpleRecorderFunc,
-	}
-
-	start := time.Now()
-	r.Record("A.B", RPCTypeInternal, start, struct{}{}, false)
+func (r readReqWithTD) IsRead() bool {
+	return true
+}
 
-	expectedLabels := []metrics.Label{
-		{Name: "method", Value: "A.B"},
-		{Name: "errored", Value: "false"},
-		{Name: "request_type", Value: "unreported"},
-		{Name: "rpc_type", Value: RPCTypeInternal},
-	}
+func (r readReqWithTD) RequestDatacenter() string {
+	return "dc3"
+}
 
-	o := store.get(append(metricRPCRequest, expectedLabels[0].Value))
-	require.Equal(t, o.key, metricRPCRequest)
-	require.LessOrEqual(t, o.elapsed, float32(start.Sub(time.Now()).Milliseconds()))
-	require.Equal(t, o.labels, expectedLabels)
+func (r readReqWithTD) GetMinQueryIndex() uint64 {
+	return 1
+}
+func (r readReqWithTD) AllowStaleRead() bool {
+	return false
 }
 
-// TestRequestRecorder_ReadRequest tests that RequestRecorder can record a read request AND a responseErrored arg.
-func TestRequestRecorder_ReadRequest(t *testing.T) {
-	t.Parallel()
+func (w writeReqWithTD) IsRead() bool {
+	return false
+}
 
-	r := RequestRecorder{
-		Logger:       hclog.NewInterceptLogger(&hclog.LoggerOptions{}),
-		RecorderFunc: simpleRecorderFunc,
-	}
+func (w writeReqWithTD) RequestDatacenter() string {
+	return "dc2"
+}
 
-	start := time.Now()
+type testCase struct {
+	name string
+	// description is meant for human friendliness
+	description string
+	// requestName is encouraged to be unique across tests to
+	// avoid lock contention
+	requestName string
+	requestI    interface{}
+	rpcType     string
+	errored     bool
+	isLeader    func() bool
+	dc          string
+	// the first element in expectedLabels should be the method name
+	expectedLabels []metrics.Label
+}
 
-	r.Record("B.A", RPCTypeNetRPC, start, readRequest{}, true)
+var testCases = []testCase{
+	{
+		name:        "simple ok",
+		description: "This is a simple happy path test case. We check for pass through and normal request processing",
+		requestName: "A.B",
+		requestI:    struct{}{},
+		rpcType:     "test",
+		errored:     false,
```

Comment:
If you wanted to make this a little more terse, you could omit this `errored` field. Also, I don't see this field ever flipped to `true` and asserted that it makes its way through the recorder. Given that `false` is the zero value for `bool`, it's probably worth adding another case for it.